server/chat-plugins/battlesearch.ts

```diff
@@ -106,140 +155,211 @@ export async function runBattleSearch(userids: ID[], month: string, tierid: ID,
 
 			const outcomes = results[day].totalOutcomes;
 			if (outcomes) {
-				outcomes.push({won: winnerid, lost: loser, turns: data.turns});
+				outcomes.push({won: winnerid, lost: loser, turns: row.turns});
 			}
 			// we only want foe data for single-userid searches
-			const foe = userids.length > 1 ? null : userid === toID(data.p1) ? toID(data.p2) : toID(data.p1);
+			const foe = userids.length > 1 ? null : userids[0] === p1id ? p2id : p1id;
 			if (foe) {
 				if (!results[day].timesBattled[foe]) results[day].timesBattled[foe] = 0;
 				results[day].timesBattled[foe]++;
 			}
 		}
 		return results;
 	}
-	for (const file of files) {
-		const subFiles = FS(`${file}`).readdirSync();
-		const day = file.split('/')[3];
-		for (const dayFile of subFiles) {
-			const json = FS(`${file}/${dayFile}`).readIfExistsSync();
-			const data = JSON.parse(json);
-			const p1id = toID(data.p1);
-			const p2id = toID(data.p2);
-			if (userids.length > 1) {
-				// looking for specific userids, only register ones where those users are players
-				if (userids.filter(item => item === p1id || item === p2id).length < userids.length) continue;
-			} else {
-				if (!(p1id === userid || p2id === userid)) continue;
-			}
-			if (turnLimit && data.turns > turnLimit) continue;
-			if (!results[day]) {
-				results[day] = {
-					totalBattles: 0,
-					totalWins: {},
-					totalOutcomes: [],
-					totalLosses: {},
-					totalTies: 0,
-					timesBattled: {},
-				};
-			}
-			results[day].totalBattles++;
-			const winnerid = toID(data.winner);
-			const loser = winnerid === p1id ? p2id : p1id;
-			if (userids.includes(winnerid)) {
-				if (!results[day].totalWins[winnerid]) results[day].totalWins[winnerid] = 0;
-				results[day].totalWins[winnerid]++;
-			} else if (data.winner) {
-				if (!results[day].totalLosses[loser]) results[day].totalLosses[loser] = 0;
-				results[day].totalLosses[loser]++;
-			} else {
-				results[day].totalTies++;
-			}
-			// explicitly state 0 of stats if none
-			for (const id of userids) {
-				if (!results[day].totalLosses[id]) results[day].totalLosses[id] = 0;
-				if (!results[day].totalWins[id]) results[day].totalWins[id] = 0;
+	async listMonths() {
+		const rows = await this.database.query(
+			`SELECT date FROM battle_logs`
```

Comment:
You are selecting every row from `battle_logs`

Reply:
is there a way to only select one of each differing row?

Reply:
`select distinct`, but you need to run the query analyzer on it to see if it will be efficient enough. It may work efficiently with an index on the appropriate columns. If not, you need to make a different table with this stuff & reference it from `battle_logs`